File: LeetCode/Problem 268.py
# LeetCode: Problem 68
# Jonathan Kosasih

#Given an array containing n distinct numbers taken from 0, 1, 2, ..., n, find the one that is missing from the array.
#
#Example 1:
#Input: [3,0,1]
#Output: 2
#
#Example 2:
#Input: [9,6,4,2,3,5,7,0,1]
#Output: 8
#
#Note:
#Your algorithm should run in linear runtime complexity. Could you implement it using only constant extra space complexity?
#------------------------------------------------------------------
class Solution:
    def missingNumber(self, nums) -> int:
        if(not nums):
            return 0
        record = sum(list(range(len(nums)+1)))
        for num in nums:
            record -= num
        return record
    
    def missingNumberAlt(self, nums) -> int:
        num_set = set(nums)
        n = len(nums) + 1
        for number in range(n):
            if number not in num_set:
                return number
                
        
sol = Solution()
example = [3,0,1]
print(str(sol.missingNumber(example)))
example = [9,6,4,2,3,5,7,0,1]
print(str(sol.missingNumber(example)))

File: LeetCode/C++/Problem 268.cpp
// LeetCode: Problem 268
// Jonathan Kosasih
/*
Given an array containing n distinct numbers taken from 0, 1, 2, ..., n, find the one that is missing from the array.
*/
//-------------------------------------------------------------------

#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <set>
#include <string>
#include <algorithm> 
#include <stack>
#include <queue>
#include <deque>
#include <numeric>
using std::vector;
using std::map;
using std::unordered_map;
using std::set;
using std::stack;
using std::queue;
using std::deque;
using std::accumulate;

class Solution {
public:
    int missingNumber(vector<int>& nums) {
        int missing = nums.size();
        for (int i = 0; i < nums.size(); i++) {
            missing ^= i ^ nums[i];
        }
        return missing;
    }
};
int main()
{
    //std::cout << "Hello World!\n";
}